core/src/main/scala/akka/kafka/internal/ConsumerResetProtection.scala

```diff
@@ -0,0 +1,103 @@
+import akka.kafka.internal.KafkaConsumerActor.Internal.Seek
+import org.apache.kafka.clients.consumer.{ConsumerRecord, ConsumerRecords}
+import org.apache.kafka.common.TopicPartition
+
+import scala.jdk.CollectionConverters._
+
+trait ConsumerResetProtection {
+  def protect[K, V](records: ConsumerRecords[K, V]): ConsumerRecords[K, V]
+}
+
+object ConsumerResetProtection {
+  def apply[K, V](ref: ActorRef,
+                  log: LoggingAdapter,
+                  threshold: Long,
+                  progress: () => ConsumerProgressTracking): ConsumerResetProtection = {
+    if (threshold > 0) new Impl(ref, log, threshold, progress()) else new Noop()
+  }
+
+  private final class Noop() extends ConsumerResetProtection {
+    override def protect[K, V](records: ConsumerRecords[K, V]): ConsumerRecords[K, V] = records
+  }
+
+  private final class Impl(consumer: ActorRef,
+                           log: LoggingAdapter,
+                           allowedMaxDeltaFromRequest: Long,
+                           progress: ConsumerProgressTracking)
+      extends ConsumerResetProtection {
+    override def protect[K, V](records: ConsumerRecords[K, V]): ConsumerRecords[K, V] = {
+      // check the fetched offsets of the records for each partition are not wildly different than the
+      // records that we have seen thus far
+      val safe: java.util.Map[TopicPartition, java.util.List[ConsumerRecord[K, V]]] =
+        records
+          .partitions()
+          .asScala
+          .map(tp => {
+            val partitionRecords = records.records(tp)
+            progress.requestedOffsets.get(tp) match {
+              case Some(requested) =>
+                val previouslyRequestedOffset = requested.offset()
+                val threshold = previouslyRequestedOffset - allowedMaxDeltaFromRequest
+                // if there are records found outside the threshold:
+                // 1. Drop the records returned for the partition (they are outside the threshold)
+                // 2. Request a seek to the latest committed offset of the partition (known to be "safe").
+                // 3. New records for the partition arrive with the next poll, which should be within the threshold
+                if (recordsExceedThreshold(threshold, partitionRecords)) {
+                  // requested and committed are assumed to be kept in-sync, so this _should_ be safe. Fails
+                  // catastrophically if this is not the case
+                  val committed = progress.committedOffsets(tp)
+                  log.warning(
+                    s"Dropping offsets for partition $tp - received an offset which is less than "
+                    + s"allowed $threshold from the last requested offset "
+                    + s"(threshold: $allowedMaxDeltaFromRequest). "
+                    + s"Seeking to the latest known safe (committed or assigned) offset: $committed"
+                  )
+                  consumer ! Seek(Map(tp -> committed.offset()))
```

Comment:
This is the central piece. I think it deserves a method.

Would it be better to pass the actor reference to `protect` instead of the instance value? (Comes a bit as a surprise that it sends messages back to the actor that called it.)

Reply:
I went back on forth on that API - didn't have a good feel on style for the actor refs, can do!